### PR TITLE
147178 - BS ancillary files delete product update serial

### PR DESCRIPTION
**Pull request by chaital14503**
This PR consists changes of [PBI 147178](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/147178)

Code changes to update SERIAL.ENC based on Type and delete PRODUCTS.TXT and INFO folder, and Unit test cases

### CODE REVIEW

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs [hunk 1]
```diff
@@ -708,7 +708,27 @@ public enum EventIds
...
+
+        /// <summary>
+        /// 809143 - SERIAL.ENC file update operation failed
+        /// </summary>
+        SerialEncUpdateFailed = 809143,
+
+        /// <summary>
+        /// 809144 - PRODUCT.TXT file and INFO folder delete operation failed
+        /// </summary>
+        ProductTxtDeleteFailed = 809144
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs:731]:
```suggestion
        BessProductTxtAndInfoFolderDeleteFailed = 809144
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs [hunk 2]
```diff
@@ -28,6 +28,14 @@ public interface IFileSystemHelper
 
         void CreateZipFile(string sourceDirectoryName, string destinationArchiveFileName, bool deleteOldArchive = false);
 
+
+        string ReadFileText(string filePath);
+
+        bool CreateFileContent(string filePath, string content);
+
+        void DeleteFile(string file);
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs:36]:
```suggestion
        void DeleteFile(string filePath);
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs [hunk 3]
```diff
@@ -178,5 +178,31 @@ public bool DownloadReadmeFile(string filePath, Stream fileStream)
...
+            return _fileSystem.File.ReadAllText(filePath);
+        }
+
+        public bool CreateFileContent(string filePath, string content)
+        {
+            if (string.IsNullOrWhiteSpace(content))
+            {
+                return false;
+            }
+            _fileSystem.File.WriteAllText(filePath, content);
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs:193]:
```suggestion
      if (_fileSystem.Directory.Exists(filePath))
      {
            _fileSystem.File.WriteAllText(filePath, content);
      }
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs [hunk 4]
```diff
@@ -178,5 +178,31 @@ public bool DownloadReadmeFile(string filePath, Stream fileStream)
             fileStream.CopyTo(outputFileStream);
             return true;
         }
+
+        public string ReadFileText(string filePath)
+        {
+            return _fileSystem.File.ReadAllText(filePath);
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs:184]:
```suggestion
         if (_fileSystem.Directory.Exists(filePath))
         {
            return _fileSystem.File.ReadAllText(filePath);
         }
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs [hunk 5]
```diff
@@ -178,5 +178,31 @@ public bool DownloadReadmeFile(string filePath, Stream fileStream)
...
+                return false;
+            }
+            _fileSystem.File.WriteAllText(filePath, content);
+
+            return true;
+        }
+
+        public void DeleteFile(string filePath)
+        {
+            _fileSystem.File.Delete(filePath);
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs:200]:
```suggestion
         if (_fileSystem.Directory.Exists(filePath))
         {
            _fileSystem.File.Delete(filePath);
         }
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs [hunk 6]
```diff
@@ -178,5 +178,31 @@ public bool DownloadReadmeFile(string filePath, Stream fileStream)
...
+        }
+
+        public void DeleteFolder(string folderPath)
+        {
+            _fileSystem.Directory.Delete(folderPath);
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs:205]:
```suggestion
          if (_fileSystem.Directory.Exists(folderPath))
          {
            _fileSystem.Directory.Delete(folderPath);
          }
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 7]
```diff
@@ -301,11 +303,20 @@ private void LogProductVersions(ProductVersionsRequest productVersionsRequest, s
...
         }
 
-        private async Task PerformAncillaryFilesOperationsAsync(string batchId, string exchangeSetPath, string correlationId, string readMeSearchFilter)
+        private async Task PerformAncillaryFilesOperationsAsync(string batchId, ConfigQueueMessage configQueueMessage, string essFileDownloadPath)
         {
+            string exchangeSetPath = Path.Combine(homeDirectoryPath, batchId, fssApiConfig.Value.BespokeExchangeSetFileFolder);
             string exchangeSetRootPath = Path.Combine(exchangeSetPath, fssApiConfig.Value.EncRoot);
             string readMeFilePath = Path.Combine(exchangeSetRootPath, fssApiConfig.Value.ReadMeFileName);
-            await CreateReadMeFileAsync(batchId, correlationId, readMeSearchFilter, exchangeSetRootPath, readMeFilePath);
+            await CreateReadMeFileAsync(batchId, configQueueMessage.CorrelationId, configQueueMessage.ReadMeSearchFilter, exchangeSetRootPath, readMeFilePath);
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs:311]: this method can be moved down above UpdateSerialFileAsync method. 
@chaital14503 pls make these changes in PBI - 140041 or 154171

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 8]
```diff
@@ -345,5 +356,49 @@ private async Task<bool> DownloadReadMeFileAsync(string batchId, string exchange
...
             return isDownloadReadMeFileSuccess;
         }
+
+        private async Task UpdateSerialFileAsync(string serialFilePath, string exchangeSetType, string correlationId)
+        {
+            string serialFileContent = String.Empty;
+            const string searchText = "UPDATE";
+            try
+            {
+                serialFileContent = fileSystemHelper.ReadFileText(serialFilePath);
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs:366]: can declared closer to usage
```suggestion
            const string searchText = "UPDATE";
            try
            {
                string serialFileContent = fileSystemHelper.ReadFileText(serialFilePath);
```

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 9]
```diff
@@ -345,5 +356,49 @@ private async Task<bool> DownloadReadMeFileAsync(string batchId, string exchange
...
+
+                    fileSystemHelper.CreateFileContent(serialFilePath, serialFileContent);
+
+                    logger.LogInformation(EventIds.BessSerialEncUpdated.ToEventId(), "SERIAL.ENC file updated with Type: {exchangeSetType} | _X-Correlation-ID:{CorrelationId}", exchangeSetType, correlationId);
+                }
+            }
+            catch (Exception ex)
+            {
+                logger.LogError(EventIds.SerialEncUpdateFailed.ToEventId(), "SERIAL.ENC file update operation failed at {DateTime} | {ErrorMessage} | _X-Correlation-ID:{CorrelationId}", DateTime.Now.ToUniversalTime(), ex.Message, CommonHelper.CorrelationID);
+                throw new Exception($"SERIAL.ENC file update operation failed at {DateTime.UtcNow} | _X-Correlation-ID:{CommonHelper.CorrelationID}", ex);
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs:379]:
```suggestion
                throw new FulfilmentException(EventIds.SerialEncUpdateFailed.ToEventId());
```

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 10]
```diff
@@ -345,5 +356,49 @@ private async Task<bool> DownloadReadMeFileAsync(string batchId, string exchange
...
+                fileSystemHelper.DeleteFile(productFilePath);
+
+                fileSystemHelper.DeleteFolder(infoFolderPath);
+
+                logger.LogInformation(EventIds.BessProductTxtAndInfoFolderDeleted.ToEventId(), "PRODUCT.TXT file and INFO folder deleted | _X-Correlation-ID:{CorrelationId}", correlationId);
+            }
+            catch (Exception ex)
+            {
+                logger.LogError(EventIds.ProductTxtDeleteFailed.ToEventId(), "PRODUCT.TXT file and INFO folder delete operation failed at {DateTime} | {ErrorMessage} | _X-Correlation-ID:{CorrelationId}", DateTime.Now.ToUniversalTime(), ex.Message, CommonHelper.CorrelationID);
+                throw new Exception($"PRODUCT.TXT file and INFO folder delete operation failed at {DateTime.UtcNow} | _X-Correlation-ID:{CommonHelper.CorrelationID}", ex);
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs:398]:
```suggestion
throw new FulfilmentException(EventIds.ProductTxtDeleteFailed.ToEventId());
```

**Review by Nadeem14456 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs (hunk 1 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs (hunk 3 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs:197]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs (hunk 5 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs (hunk 6 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs (hunk 8 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs (hunk 9 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs (hunk 7 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs:311]: Noted

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs (hunk 2 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs]: Resolved

**Review by Nadeem14456 — APPROVED** [review]
LGTM. 
Note: There are few suggestions which will be incorporated in PBI 154171.

**Review by DanOwenUKHO — APPROVED** [review]
LGTM

**Pull request merged**